### PR TITLE
CGAL 3D Demo: Replace QDoubleSpinBoxes

### LINKED ISSUE #4251: Demos:  Dialog with numbers are too short
## Issue Details

For example in the isotropic remeshing pop up window I first must remove the last digit, so that I can insert a 0 right after the comma.  

image

### PR DESCRIPTION
**Pull request by maxGimeno**
## Summary of Changes

Replace all `QDoubleSpinBox`es by a custom class deriving from QLineEdit associated with a validator to restaint it to doubles. 
## Release Management

* Issue(s) solved (if any): fix #4251

### CODE REVIEW

File: Polyhedron/demo/Polyhedron/CGAL_double_edit.cpp [hunk 1]
```diff
@@ -0,0 +1,43 @@
+#include "CGAL_double_edit.h"
+
+#include <QDoubleValidator>
+
+  DoubleEdit::DoubleEdit(QWidget *parent)
+    : QLineEdit()
```
Comment (lrineau) [inline Polyhedron/demo/Polyhedron/CGAL_double_edit.cpp]:
```suggestion
    : QLineEdit(parent)
```

File: Polyhedron/demo/Polyhedron/CGAL_double_edit.h [hunk 2]
```diff
@@ -0,0 +1,31 @@
+#ifndef CGAL_DOUBLE_EDIT_H
+#define CGAL_DOUBLE_EDIT_H
+
+
+
+#include <QObject>
+#include <QtCore/qglobal.h>
+#include <QLineEdit>
+#ifdef cgal_double_edit_EXPORTS
```
Comment (lrineau) [inline Polyhedron/demo/Polyhedron/CGAL_double_edit.h]:
```suggestion
#ifdef demo_framework_EXPORTS
```

File: Polyhedron/demo/Polyhedron/Plugins/Classification/Classification_plugin.cpp [hunk 3]
```diff
@@ -619,17 +619,14 @@ public Q_SLOTS:
...
-    voxel_size->setSingleStep (0.01);
+    DoubleEdit* voxel_size = dialog.add<DoubleEdit> ("Voxel size (0 for automatic):");
 
     if (dialog.exec() != QDialog::Accepted)
       return;
 
     QApplication::setOverrideCursor(Qt::WaitCursor);
 
-    float vsize = float(voxel_size->value());
+    float vsize = float(voxel_size->text().toDouble());
```
Comment (lrineau) [inline Polyhedron/demo/Polyhedron/Plugins/Classification/Classification_plugin.cpp]: Why this change? It seems your new class also has `value()`.

File: Polyhedron/demo/Polyhedron/Plugins/Classification/Classification_plugin.cpp [hunk 4]
```diff
@@ -921,18 +918,17 @@ public Q_SLOTS:
...
+    smoothing->setText(tr("%1").arg(0.5));
 
     if (dialog.exec() != QDialog::Accepted)
       return;
     
     QApplication::setOverrideCursor(Qt::WaitCursor);
     CGAL::Real_timer t;
     t.start();
-    run (classif, 2, std::size_t(subdivisions->value()), smoothing->value());
+    run (classif, 2, std::size_t(subdivisions->value()), smoothing->text().toDouble());
```
Comment (lrineau) [inline Polyhedron/demo/Polyhedron/Plugins/Classification/Classification_plugin.cpp]: `value()` is fine.

File: Polyhedron/demo/Polyhedron/Plugins/Classification/Classification_plugin.cpp [hunk 5]
```diff
@@ -1371,10 +1367,8 @@ public Q_SLOTS:
       QSpinBox* trials = dialog.add<QSpinBox> ("Number of trials: ", "trials");
       trials->setRange (1, 99999);
       trials->setValue (500);
-      QDoubleSpinBox* rate = dialog.add<QDoubleSpinBox> ("Learning rate: ", "learning_rate");
-      rate->setRange (0.00001, 10000.0);
-      rate->setValue (0.001);
-      rate->setDecimals (5);
```
Comment (lrineau) [inline Polyhedron/demo/Polyhedron/Plugins/Classification/Classification_plugin.cpp]: I wonder why you change that. Given that your new class has `setRange` and `setValue`.

File: Polyhedron/demo/Polyhedron/Plugins/Display/Display_property.ui [hunk 6]
```diff
@@ -110,44 +110,18 @@
...
            <item>
-            <widget class="QDoubleSpinBox" name="minBox"/>
-           </item>
-           <item>
-            <widget class="QDoubleSpinBox" name="maxBox">
-             <property name="value">
-              <double>2.000000000000000</double>
+            <widget class="DoubleEdit" name="minBox">
+             <property name="text">
+              <string>0,00</string>
```
Comment (lrineau) [inline Polyhedron/demo/Polyhedron/Plugins/Display/Display_property.ui]: `0,00`?! You set the locale to `C`. It should be `0.00`.

File: Polyhedron/demo/Polyhedron/Plugins/Display/Display_property.ui [hunk 7]
```diff
@@ -110,44 +110,18 @@
...
-               </property>
-               <property name="autoRepeat">
-                <bool>false</bool>
-               </property>
-              </widget>
-             </item>
-            </layout>
+            <widget class="DoubleEdit" name="maxBox">
+             <property name="text">
+              <string>2,00</string>
```
Comment (lrineau) [inline Polyhedron/demo/Polyhedron/Plugins/Display/Display_property.ui]: `2.0`

File: Polyhedron/demo/Polyhedron/Plugins/Display/Display_property_plugin.cpp [hunk 8]
```diff
@@ -443,9 +443,6 @@ class DisplayPropertyPlugin :
     connect(dock_widget->deleteButton, &QPushButton::clicked,
             this, &DisplayPropertyPlugin::delete_group);
 
-    connect(dock_widget->resetButton, &QPushButton::pressed,
-            this, &DisplayPropertyPlugin::resetRampExtremas);
```
Comment (lrineau) [inline Polyhedron/demo/Polyhedron/Plugins/Display/Display_property_plugin.cpp]: That does not seem related to spin boxes.

File: Polyhedron/demo/Polyhedron/Plugins/Mesh_3/Meshing_dialog.ui [hunk 9]
```diff
@@ -186,6 +157,20 @@
         </property>
        </widget>
       </item>
+      <item row="3" column="1">
+       <widget class="DoubleEdit" name="sharpEdgesAngle">
+        <property name="text">
+         <string>60,00</string>
```
Comment (lrineau) [inline Polyhedron/demo/Polyhedron/Plugins/Mesh_3/Meshing_dialog.ui]: Again: `60` instead of `60,00`.

File: Polyhedron/demo/Polyhedron/Plugins/Mesh_3/Meshing_dialog.ui [hunk 10]
```diff
@@ -186,6 +157,20 @@
...
+        </property>
+       </widget>
+      </item>
+      <item row="7" column="1">
+       <widget class="DoubleEdit" name="edgeSizing">
+        <property name="text">
+         <string>0,0</string>
```
Comment (lrineau) [inline Polyhedron/demo/Polyhedron/Plugins/Mesh_3/Meshing_dialog.ui]: `0.0`. It do not understand that error. I probably miss something.

File: Polyhedron/demo/Polyhedron/Plugins/Mesh_3/Meshing_dialog.ui [hunk 11]
```diff
@@ -359,27 +315,25 @@
...
-         <double>0.000000000000000</double>
+       <widget class="DoubleEdit" name="facetAngle">
+        <property name="inputMask">
+         <string>25,00</string>
         </property>
-        <property name="maximum">
-         <double>30.000000000000000</double>
-        </property>
-        <property name="value">
-         <double>25.000000000000000</double>
```
Comment (lrineau) [inline Polyhedron/demo/Polyhedron/Plugins/Mesh_3/Meshing_dialog.ui]: There was a meaning in the minimum and maximum values. I think the `QDoubleValidator` can implement that.

I stop my review here. There are similar issues below.

File: Polyhedron/demo/Polyhedron/Plugins/Classification/Classification_plugin.cpp (hunk 5 above)
Comment (maxGimeno) [reply Polyhedron/demo/Polyhedron/Plugins/Classification/Classification_plugin.cpp]: At first I didn't think to code those functions, I did it when I realized there were FAR MORE QDoubleSpinBoxes than I thought and it became clear that adding those functions would be a lot faster than to remove the calls everywhere. So the first classes I worked on were managed differently.

File: Polyhedron/demo/Polyhedron/Plugins/Display/Display_property_plugin.cpp (hunk 8 above)
Comment (maxGimeno) [reply Polyhedron/demo/Polyhedron/Plugins/Display/Display_property_plugin.cpp]: It is not, the commit explains that change. I spotted a bug while testing the spin box replacement, and "fixed " it here.

File: Polyhedron/demo/Polyhedron/CGAL_double_edit.h [hunk 12]
```diff
@@ -0,0 +1,28 @@
...
+
+  Q_OBJECT
+
+public:
+  DoubleEdit(QWidget* parent = nullptr);
+  ~DoubleEdit();
+  double value() const;
+  void setValue(double d);
+  void setMinimum(double d);
+  void setMaximum(double d);
```
Comment (lrineau) [inline Polyhedron/demo/Polyhedron/CGAL_double_edit.h]: `minimum`, `maximum`, and `value` should be declared as properties (using `Q_PROPERTY`). I think that allows you to keep the compatibility with the properties from the UI file.

**Comment by maxGimeno** [comment]
https://cgal.geometryfactory.com/CGAL/testsuite/results-5.0-Ic-173.shtml

**Comment by maxGimeno** [comment]
Successfully tested in https://cgal.geometryfactory.com/CGAL/testsuite/results-5.1-Ic-18.shtml
